### PR TITLE
DS-2698: Add badge playwright e2e tests

### PR DESCRIPTION
**Pull request by tresajm**
### **Summary**
The ontario-badge.e2e.ts file contains end-to-end (E2E) tests for the Ontario Badge component, validating its rendering, styling, accessibility and state changes using Playwright-driven browser testing..
### **Purpose**
This test suite ensures the Ontario Badge component remains:
- Functionally correct
- Accessible and standards-compliant
- Visually consistent across supported variants
- Backward compatible with legacy inputs
- Stable under high-frequency updates and rendering operations

The file provides comprehensive coverage of expected usage patterns, edge cases, accessibility requirements, and performance considerations for the badge component.

### Test Coverage
**Positive Tests**
- Verifies that the badge renders correctly and hydrates successfully.
- Confirms label text is displayed when provided.
- Validates slot content rendering when no label property is supplied.
- Ensures the correct typography styles are applied to badge labels.
- Checks that badge colour variants (e.g., red) render with expected styles.
- Verifies colour changes are reflected when updated dynamically.
- Confirms aria-label values are applied correctly for accessibility.

**Negative Tests**
- Ensures invalid colour values gracefully fall back to the default teal colour.
- Verifies the component renders appropriately when neither label nor slot content is provided.
- Tests handling of empty aria-label values.
- Validates backward compatibility by mapping legacy colour values to supported variants.

**Boundary and Edge Case Tests**
- Tests handling of labels longer than 15 characters.
- Verifies behaviour when an empty string is provided as a label.
- Ensures the default teal colour is applied when no colour is specified.
- Tests component stability during rapid property updates.

**Accessibility Tests**
- Confirms the component has no accessibility violations.
- Verifies screen readers use the provided aria-label.
- Ensures visible text is used as an accessible fallback when aria-label is absent.
- Validates semantic HTML structure and accessibility compliance.

**Visual Regression Tests**
- Includes visual regression test coverage for badge colour variants.
- Currently skipped due to limitations with Playwright screenshot support in Stencil E2E testing.

**Performance Tests**
- Verifies successful rendering of multiple badge instances.
- Evaluates component behaviour during rapid re-rendering scenarios.

### CODE REVIEW

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 1]
```diff
@@ -1,105 +1,263 @@
...
+		await host.evaluate((el: any) => {
+			el.colour = 'alert';
+			el.colour = 'teal';
+			el.colour = 'alert';
+		});
+
+		await page.waitForChanges();
+
+		await expect(host.locator('span')).toBeAttached();
+	});
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:165]: This test looks like it is testing rapid changes to the `colour` property, however it isn't validating the expected result from the changes.  Let's improve the validation in the test to it is more robust.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 2]
```diff
@@ -1,105 +1,263 @@
...
+		const host = await renderHost(page, `<ontario-badge label="Default"></ontario-badge>`);
+
+		await expect(host.locator('span')).toContainClass('ontario-badge--teal');
+	});
+
+	test('handles rapid updates', async ({ page }) => {
+		const host = await renderHost(page, `<ontario-badge label="Stress"></ontario-badge>`);
+
+		await host.evaluate((el: any) => {
+			el.colour = 'alert';
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: `alert` is not one of our valid colours, is this intentional here?

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 3]
```diff
@@ -1,105 +1,263 @@
...
+
+	/* =========================
+     Visual Regression Tests
+    ========================== */
+
+	test('visual regression: default badge', async ({ page }) => {
+		const host = await renderHost(page, `<ontario-badge label="Snapshot"></ontario-badge>`);
+
+		const screenshot = await host.screenshot();
+		expect(screenshot.byteLength).toBeGreaterThan(0);
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: This is currently checking the screenshot was created, however, we should be checking the contents of the screenshots match.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 4]
```diff
@@ -1,105 +1,263 @@
...
-// 	});
+	const renderHost = async (page: E2EPage, html: string): Promise<Locator> => {
+		await page.setContent(html);
+		await page.waitForChanges();
 
-// 	describe('render prop changes', () => {
-// 		let page: E2EPage;
-// 		let component: E2EElement;
-// 		let element: E2EElement;
+		const host = page.locator('ontario-badge').last();
```
Comment (muntalee) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:14]: This helper always returns the last badge on the page. That makes single-badge tests fine, but any test that renders multiple badges will only assert against the final instance. Maybe we can return a more explicit locator here, or have the multi-badge tests select each badge directly?

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 5]
```diff
@@ -1,105 +1,263 @@
...
+
+		const screenshot = await host.screenshot();
+		expect(screenshot.byteLength).toBeGreaterThan(0);
+	});
+
+	test('visual regression: colour variants', async ({ page }) => {
+		const host = await renderHost(
+			page,
+			`
+      <ontario-badge label="One" colour="teal"></ontario-badge>
```
Comment (muntalee) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: This test says it is covering colour variants, but because renderHost returns only the last badge, it only screenshots the second badge. Also, is `alert` a valid badge colour?

**Review by muntalee — CHANGES_REQUESTED** [review]
Hey Tresa! Here's some reviews

**Review by smorrisods — CHANGES_REQUESTED** [review]
There is currently no description on this pull request describing the changes. Please update this prior to it being ready to merge.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 4 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:14]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 5 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 1 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:165]: validation is added

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 2 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: alert is removed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 3 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: updated with the change requested

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 6]
```diff
@@ -1,105 +1,302 @@
...
+      <ontario-badge label="Two" colour="green"></ontario-badge>
+      `,
+		);
+
+		await expect(host).toHaveScreenshot('colour-variants.png');
+	});
+
+	/* =========================
+     Performance Tests
```
Comment (erincarrieresontario) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:210]: I think this is less assessing performance and more assessing stability

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 7]
```diff
@@ -1,105 +1,302 @@
...
 
-// 		beforeEach(async () => {
-// 			page = await newE2EPage();
-// 			await page.setContent('<ontario-badge></ontario-badge>');
-// 			component = await page.find('ontario-badge');
-// 			element = await page.find('ontario-badge >>> span');
-// 		});
+		// Ensure the page has finished loading resources and fonts are ready
+		try {
+			await page.waitForLoadState('networkidle');
```
Comment (erincarrieresontario) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Consider whether networkidle is needed here. For component-level tests, waiting for component rendering (waitForChanges) and font readiness may be sufficient and could reduce test runtime.

**Review by erincarrieresontario — COMMENTED** [review]
Looks good @tresajm , just a few comments. Also there are merge conflicts.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 7 above)
Comment (tresavarghese1) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: updated

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 8]
```diff
@@ -1,105 +1,292 @@
...
+		await expect(span).toBeAttached();
+		await expect(span).not.toHaveAttribute('role', /.+/);
+		await expect(host).toHaveScreenshot('semantic-structure.png');
+	});
+
+	/* =========================
+     Visual Regression Tests -- currently skipped due to Playwright screenshot not implemented in Stencil E2E testing
+    ========================== */
+
+	test('visual regression: colour variants', async ({ page }) => {
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: This test renders **two** badges (`teal` "One" and `green` "Two") but takes its screenshot via `host`, which comes from `renderHost()`. Tracing through `renderHost` at line 14, `host` is always `page.locator('ontario-badge').last()` — so the committed `colour-variants-chromium.png` only ever captures the second (`green` "Two") badge, not both colour variants.

I pulled the committed snapshot and confirmed it: it's a 55x17px image showing just the green "Two" badge, nothing from the teal one.

This is actually the same root cause muntalee flagged earlier in this thread (`renderHost` always returning `.last()`), marked resolved with "fixed" — but for this specific test the fix doesn't hold up, since it still routes through `renderHost` and only gets the last badge.

```suggestion
	test('visual regression: colour variants', async ({ page }) => {
		await page.setContent(`
      <ontario-badge label="One" colour="teal"></ontario-badge>
      <ontario-badge label="Two" colour="green"></ontario-badge>
      `);
		await page.waitForChanges();

		await expect(page).toHaveScreenshot('colour-variants.png');
	});
```

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 9]
```diff
@@ -1,105 +1,292 @@
...
+		// Check class mapping for alert colour
+		await expect(span).toContainClass('ontario-badge--red');
+		await expect(host.locator('span')).toHaveScreenshot('red-badge.png');
+	});
 
-// 		it('renders changes to the component content when the host textContent is changed', async () => {
-// 			await page.setContent('<ontario-badge>Not started</ontario-badge>');
-// 			const element = await page.find('ontario-badge >>> span');
+	test('updates colour dynamically', async ({ page }) => {
+		const host = await renderHost(page, `<ontario-badge label="Dynamic"></ontario-badge>`);
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:68]: `el: any` shows up here and again at lines 165 and 277 for `host.evaluate()` callbacks. We've already got a generated `HTMLOntarioBadgeElement` type (`components.d.ts`) and other component e2e tests (`ontario-radio-buttons`, `ontario-textarea`) type their `evaluate` callbacks against the matching `HTMLOntario*Element` type instead of `any`. Let's do the same here so `el.colour` stays type-checked against `BadgeColour`:

```suggestion
		await host.evaluate((el: HTMLOntarioBadgeElement) => {
```

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 10]
```diff
@@ -1,105 +1,292 @@
...
+		await expect(host.locator('span')).toHaveScreenshot('unicode-badge.png');
+	});
+
+	/* =========================
+     Accessibility Tests
+    ========================== */
+
+	test('has no accessibility violations', async ({ page }) => {
+		const host = await renderHost(page, `<ontario-badge label="Accessible"></ontario-badge>`);
+		await expectNoAxeViolations(page, 'ontario-badge');
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:179]: `expectNoAxeViolations(page, 'ontario-badge')` on the line above already asserts `results.violations` has length 0 (see the helper at line 21-25) — so if this line passes, the manual `AxeBuilder(...).analyze()` + violation-logging block below it (lines 192-206) is dead code; the `if (results.violations.length > 0)` branch can never be entered. We can drop the duplicate scan and just keep the helper call plus the visibility check.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 11]
```diff
@@ -1,105 +1,292 @@
...
+		const host = await renderHost(page, `<ontario-badge label="Semantic"></ontario-badge>`);
+
+		const span = host.locator('span');
+		await expect(span).toBeAttached();
+		await expect(span).not.toHaveAttribute('role', /.+/);
+		await expect(host).toHaveScreenshot('semantic-structure.png');
+	});
+
+	/* =========================
+     Visual Regression Tests -- currently skipped due to Playwright screenshot not implemented in Stencil E2E testing
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: This comment says the visual regression tests are "currently skipped due to Playwright screenshot not implemented in Stencil E2E testing," but none of the tests under this section (or in the file generally, aside from the one intentional `test.skip` at line 129) are actually skipped — most `toHaveScreenshot` calls run. Let's update or remove this comment so it doesn't mislead a future reader into thinking VRT coverage isn't running.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 12]
```diff
@@ -1,105 +1,292 @@
...
+			for (let i = 0; i < 10; i++) {
+				el.colour = i % 2 ? 'red' : 'teal';
+			}
+		});
+
+		const start = performance.now();
+		await page.waitForChanges();
+
+		const end = performance.now();
+		console.log(`Rendered 50 badges in ${(end - start).toFixed(2)} ms`);
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: This log message says `Rendered 50 badges in ...ms`, but this test (`handles rapid re-rendering`) only renders a single badge (`Perf`) and mutates its `colour` 10 times — looks like a copy-paste leftover from the `renders multiple badges` test above. Let's fix the message to describe what's actually happening here.

```suggestion
		console.log(`Completed 10 rapid colour updates in ${(end - start).toFixed(2)} ms`);
```

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 13]
```diff
@@ -1,105 +1,292 @@
...
+		await expect(label).toHaveText('Fallback');
+		await expect(host).toBeVisible();
+	});
+
+	test('has correct semantic structure', async ({ page }) => {
+		const host = await renderHost(page, `<ontario-badge label="Semantic"></ontario-badge>`);
+
+		const span = host.locator('span');
+		await expect(span).toBeAttached();
+		await expect(span).not.toHaveAttribute('role', /.+/);
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Minor: `.not.toHaveAttribute('role', /.+/)` reads as "doesn't have a `role` attribute matching any non-empty value," which is a slightly roundabout way to assert the attribute isn't present at all. Would `.not.toHaveAttribute('role')` say the same thing more plainly here?

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 14]
```diff
@@ -1,105 +1,292 @@
...
+
+		await expect(host.locator('span')).toContainClass('ontario-badge--light-teal');
+		await expect(host.locator('span')).toHaveScreenshot('legacy-colour-mapping.png');
+	});
+
+	/* ==============================================
+     Boundary Tests with longer than 15 characters
+    ================================================= */
+	/* ==============================================
+    Curretly skipped because the badge component is not designed to handle long text
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Small typo: `Curretly` → `Currently`.

**Review by smorrisods — CHANGES_REQUESTED** [review]
Good work on the coverage here, @tresajm — this is a solid pass at real Playwright e2e tests for `ontario-badge` compared to the fully-commented-out stub this replaces. I did find one real gap in the visual regression coverage though (the `colour variants` screenshot only ever captures one of the two badges due to `renderHost`'s `.last()` behaviour), plus a handful of smaller nits: an `any` type we can tighten, some dead code in the axe test, a stale comment, a copy-paste log message, and a couple of small wording things.

---
🤖 Drafted by an AI running on Scott's greatest-hits review patterns. Bugs are real, banter is simulated.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 8 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 9 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:68]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 10 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:179]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 11 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 12 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: applied the change

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 13 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 14 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:1]: Visual Regression Test should be in a file ending in `.vrt.ts`, these run in an different job runner and it keeps them separate from the interactive E2E tests.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 15]
```diff
@@ -1,105 +1,277 @@
...
-// 		/*
-// 		 * TODO: Haven't found an ideal way yet to test the aria-label-text property and how
-// 		 * it should create an aria-label attribute on the element.
-// 		 */
-// 		// it('renders changes to the aria-label-text property', async () => {
-// 		// component.setProperty('aria-label-text', 'This is aria label text.');
-// 		// await page.waitForChanges();
+		await expect(host).toBeAttached();
+		await expect(host).toHaveClass('hydrated');
+		await expect(host).toHaveScreenshot('default-badge.png');
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Let's use the autogenerated image names rather than having to maintain handwritten names across many test suites.

This applies to all the `toHaveScreenshot` calls.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 16]
```diff
@@ -1,105 +1,277 @@
...
+		const host = await renderHost(page, `<ontario-badge label="Default"></ontario-badge>`);
+
+		await expect(host.locator('span')).toContainClass('ontario-badge--teal');
+		await expect(host).toHaveScreenshot('default-colour.png');
+	});
+
+	test('handles rapid updates', async ({ page }) => {
+		const host = await renderHost(page, `<ontario-badge label="Stress"></ontario-badge>`);
+
+		await host.evaluate((el: any) => {
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Let's update this to be a real type and not a generic `any` type.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 17]
```diff
@@ -1,105 +1,277 @@
...
+
+		const badges = page.locator('ontario-badge');
+		await expect(badges).toHaveCount(50);
+		await expect(page).toHaveScreenshot('multiple-badges.png');
+	});
+
+	test('handles rapid re-rendering', async ({ page }) => {
+		const host = await renderHost(page, `<ontario-badge label="Perf"></ontario-badge>`);
+
+		await host.evaluate((el: any) => {
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Let's update this to be a real type and not a generic `any` type.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 18]
```diff
@@ -1,105 +1,277 @@
...
+		);
+		await page.waitForChanges();
+
+		const container = page.locator('#colour-variants');
+		const badges = container.locator('ontario-badge');
+
+		await expect(badges).toHaveCount(2);
+		await expect(badges.last()).toHaveClass(/hydrated/);
+
+		await expect(container).toHaveScreenshot('colour-variants.png');
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: I see the entire width in the screenshot is coming from the `div` container's width rendering as the full page width. I don't think is inherently an issue, it just captures more data than is needed.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 19]
```diff
@@ -1,105 +1,277 @@
...
+	test('renders multiple badges', async ({ page }) => {
+		const html = Array.from({ length: 50 })
+			.map((_, i) => `<ontario-badge label="Badge ${i}"></ontario-badge>`)
+			.join('');
+
+		await page.setContent(html);
+		await page.waitForChanges();
+
+		const badges = page.locator('ontario-badge');
+		await expect(badges).toHaveCount(50);
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: I assume this number relates to `length: 50`.  Let's name this constant so there is less chance of drift between the test setup and the test assertion.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 20]
```diff
@@ -1,105 +1,277 @@
...
+		await expect(page).toHaveScreenshot('multiple-badges.png');
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Same idea as one of the other tests, the additional space is taken up by the container filling the page width and height.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:1]: There are only _chromium_ screenshots committed here, but we generally test against Firefox and Webkit, I wonder what is up when they rendered 🤔.

**Review by smorrisods — CHANGES_REQUESTED** [review]
The tests are coming along nicely, I've added some comments around file names, value constancy, and image sizes.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 17 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 16 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: fixed

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 15 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: removed the handwritten names

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 18 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Updated the below section
Screenshot 2026-08-11 at 12 48 36 PM

This is the screenshot
Screenshot 2026-08-11 at 12 49 11 PM

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 20 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Updated the test with 
await page.setContent(`<div id="multiple-badges" style="display: inline-flex; flex-wrap: wrap;">${html}</div>`);

The screenshot
Screenshot 2026-08-11 at 1 10 28 PM

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:1]: After creating the ontario-badge.vrt.ts file, the screenshots are captured Chromium, Firefox and Webkit

Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:1]: Added ontario-badge.vrt.ts file

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 19 above)
Comment (tresajm) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts]: Updated as mentioned in the comment above.
Screenshot 2026-08-11 at 2 29 16 PM

File: .github/workflows/build.yml
Comment (smorrisods) [inline .github/workflows/build.yml:1]: This file diff is undoing the changes we just added to the CI, it needs to be reverted back to what is in `develop` and shouldn't be modified in this PR.

File: .github/workflows/main.yml
Comment (smorrisods) [inline .github/workflows/main.yml:1]: Likewise, this file should be reverted and untouched in this PR.

File: .github/workflows/test.yml
Comment (smorrisods) [inline .github/workflows/test.yml:1]: Likewise, this file should be reverted and untouched in this PR.

File: packages/app-angular/package.json
Comment (smorrisods) [inline packages/app-angular/package.json:1]: This file also contains regressions to what is on `develop`. This file should be reverted and untouched in this PR.

File: pnpm-workspace.yaml
Comment (smorrisods) [inline pnpm-workspace.yaml:1]: This file should be reverted to what is on `develop` and untouched in this PR.

File: scripts/start-server.sh
Comment (smorrisods) [inline scripts/start-server.sh:1]: This file should be reverted to what is on `develop` and untouched in this PR.

File: patches/conventional-changelog-angular@8.0.0.patch
Comment (smorrisods) [inline patches/conventional-changelog-angular@8.0.0.patch:1]: This file should be reverted to what is on `develop` and untouched in this PR.

File: patches/@semantic-release__github@12.0.9.patch
Comment (smorrisods) [inline patches/@semantic-release__github@12.0.9.patch:1]: This file should be reverted to what is on `develop` and untouched in this PR.

File: packages/ontario-design-system-component-library/src/index.html
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/index.html:1]: This file change is reverting features that are already on `develop`, only changes related to badges should be here, if any.

**Review by smorrisods — CHANGES_REQUESTED** [review]
I've stopped commenting about the overwritten `develop` changes that run through this current iteration of this PR because there are too many to call out.  If you'd like some assistance with fixing it let me know.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/vrt-snapshots/ontario-badge-handles-empty-string-label-1-chromium.png
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/vrt-snapshots/ontario-badge-handles-empty-string-label-1-chromium.png:1]: Is this the expected rendering behaviour of our badge when there is no label inside it? @tresajm cc/ @erincarrieresontario

**Review by smorrisods — COMMENTED** [review]
This is looking good, I have one question about the empty label case we should answer first and track if it is an issue.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/vrt-snapshots/ontario-badge-handles-empty-string-label-1-chromium.png
Comment (erincarrieresontario) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/vrt-snapshots/ontario-badge-handles-empty-string-label-1-chromium.png]: @smorrisods - yes, there will just be an empty badge. Do we want a default label?

**Review by muntalee — APPROVED** [review]
LGTM @tresajm 🐐

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts [hunk 21]
```diff
@@ -1,105 +1,244 @@
...
+		await page.setContent(`<div id="multiple-badges" style="display: inline-flex; flex-wrap: wrap;">${html}</div>`);
+		await page.waitForChanges();
+
+		const container = page.locator('#multiple-badges');
+		const badges = container.locator('ontario-badge');
+		await expect(badges).toHaveCount(badgeCount);
+	});
+
+	test('handles rapid re-rendering', async ({ page }) => {
+		const host = await renderHost(page, `<ontario-badge label="Perf"></ontario-badge>`);
```
Comment (smorrisods) [inline packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:228]: This test times 10 rapid `colour` mutations with `performance.now()` and `console.log`s the result, but never asserts anything about that duration — the only real assertions (`toBeAttached`, class checks) are identical to what `'handles rapid updates'` already covers without the timing. As written, this doesn't actually test "performance," it just adds console noise to the test run.

If we want a real performance guardrail, let's assert an upper bound:
```ts
const start = performance.now();
await page.waitForChanges();
const durationMs = performance.now() - start;
expect(durationMs).toBeLessThan(500); // tune this threshold
```
Otherwise, I'd drop the timing/`console.log` and let `'handles rapid updates'` cover this case on its own. Same pattern shows up in `ontario-badge.vrt.ts`'s copy of this test too.

**Review by smorrisods — CHANGES_REQUESTED** [review]

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts (hunk 21 above)
Comment (smorrisods) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/ontario-badge.e2e.ts:228]: Good catch — fixed in 31940f8, both `ontario-badge.e2e.ts` and `ontario-badge.vrt.ts` now assert `durationMs` stays under 500ms instead of just logging it. Verified locally (chromium) that the test still passes with the real assertion in place.

File: packages/ontario-design-system-component-library/src/components/ontario-badge/test/vrt-snapshots/ontario-badge-handles-empty-string-label-1-chromium.png
Comment (smorrisods) [reply packages/ontario-design-system-component-library/src/components/ontario-badge/test/vrt-snapshots/ontario-badge-handles-empty-string-label-1-chromium.png:1]: Checked against the component source (`ontario-badge.tsx`) — this is expected. `validateLabel()` only fires a console warning when no label/slot content is provided; it doesn't hide the host or suppress rendering. So an empty `label=""` (or no label at all) intentionally still renders the coloured pill with no visible text, matching the 'renders empty when no label and no slot' e2e test. Treating this as expected behaviour — resolving, but happy to revisit if design intent differs.

**Comment by smorrisods** [comment]
Rebased onto `develop` (no conflicts) and addressed the two open review threads:

- **Rapid re-rendering test** — `ontario-badge.e2e.ts` and `ontario-badge.vrt.ts` now assert `durationMs < 500ms` instead of just `console.log`-ing the timing (31940f8).
- **Empty-label VRT question** — confirmed against `ontario-badge.tsx` that the coloured-pill-with-no-text rendering is expected behaviour (`validateLabel()` only warns, doesn't suppress rendering).

All CI checks are green (build-libs, build-apps, Stencil unit/e2e/VRT, Apps e2e/VRT). Ready for a fresh look 🙏

**Review by smorrisods — APPROVED** [review]
LGTM 🚀
